Add blog post of how contribute to template

r-gaia-cs opened the pull request:
@gvwilson, @wking, @fmichonneau and @pbanaszkiewicz Could you review this?

r-gaia-cs commented on blog/2015/04/how-to-contribute-with-template.html:
```diff
@@ -0,0 +1,112 @@
+---
+layout: blog
+root: ../../..
+author: Raniere Silva
```
Can I change this to "Mentoring Subcommittee"?

gvwilson replied:
I'd prefer to have people listed as authors - their identities tend to be more stable than those of committees.

wking commented:
Docs are good, but I'd rather keep them in the lesson-template
repository.  If you put docs in a blog post, they'll go stale, but
folks will still find them and try to follow them ;).  I'd rather land
something like this in lesson-template, and then having a “We have new
docs!” blog post that links to it.  There's also some overlap with
[1], so we might just want to incorperate this information there until
we resolve swcarpentry/lesson-template#215.

Anyhow, this run through looks good to me (and it's nice to have this
sort of thing written up).  I have a few tweaks, ranging from typo
fixes to some minor Git reshuffling, but I'd be fine if this landed
as-is for later polishing.

On Sat, Apr 04, 2015 at 05:39:58PM -0700, Raniere Silva wrote:
>   * Add blog post of how contribute to template

* “To ensure that all lessons have the same look we create…” → “To
  ensure that all lessons have the same look, we created…” (comma and
  “created”).

* “… core files of our template
  <a href="https://github.com/swcarpentry/lesson-template/pull/219">in
  one repository</a>” should maybe just be linking to lesson-template
  (not in to the 219 PR)?

* “Because create a pull request…” → “Because creating a pull
  request…”

* “…starting with Git I wrote this how-to” → “starting with Git, I
  wrote this how-to”.

* I prefer cloning over HTTPS in example docs, see [2] and later.

* It looks like you're cloning r-gaia-cs/lesson-example.  I expect
  most folks will want to clone swcarpentry/lesson-example and only
  fork that repo if they need to update the examples.  If you're just
  contributing to lesson-template, you don't need a lesson-example
  fork.

* I think novice Git users will probably be even more confused with so
  many remotes for one local repository.  It's probably easier to
  recommend folks have a separate local repository for each lesson
  they maintain as well as separate local repositories for
  lesson-template and lesson-example.  Then each local repository only
  needs remotes for the central version of that repository and the
  users personal public fork.  That's less efficient in terms of
  object de-duplication, but it makes for a simpler mental map.  More
  advanced Git users will be able to figure out their own preferred
  remote mapping without our advice.  For template work, you will need
  an extra remote pointing to a filled-in example repository
  (e.g. lesson-example), but you won't need two (either
  r-gaia-cs/swc-lesson-example *or* swcarpentry/lesson-example would
  be fine).

* `fit fetch --all` is easier than fetching each remote in turn ;).

* I prefer:

    $ git merge --ff-only upstream/gh-pages

  to avoid creating accidental backwards merges.  Alternatively, don't
  bother updating your local gh-pages and just do:

    $ git checkout -b example-centering-image upstream/gh-pages

  for “I'd like to create a feature branch about centering images
  based on upstream's gh-pages branch”.

* You have both “how-to” and “how to”.  Wikipedia prefers the
  hyphenated form [3].

* You have trailing whitespace after “git diff”.

* “To avoid lose my work I save it at the staged area” → “To avoid
  losing my work, I stage it in the index”.

* “save the change at the HTML file” → “save the change to the HTML
  file”.

* “If you attend a Software Carpentry workshop you should be
  confortable…” → “If you have taken the Git lesson at a Software
  Carpentry workshop, you should be comfortable…” (specify Git lesson
  and fix “confortable” typo).

Instead of committing the template change and HTML to the
example-based branch, I'd work out the change in the example-based
branch, then:

Blow away the side effects:

  $ git checkout HEAD *.html

Save my changes for the branch switch:

  $ git stash

Switch branches to a feature branch based on the central template:

  $ git checkout -b center-images swcarpentry-template/gh-pages

Restore my changes after the branch switch:

  $ git stash pop

Commit them:

  $ git commit -am 'Add img-responsive class to img'

Publish:

  $ git push -u my-public-fork-of-lesson-template center-images

And create a pull request:

  $ hub pull-request

(I'd probably follow you and link to the PR docs for the
instructions, but I thought I'd share my hub [4] use here, because
it makes my life a lot easier ;).

Anyhow, the stash-switch-pop-commit approach lets you avoid creating
a branch based on the example's gh-pages (you can work in a detached
HEAD if you want, or in your local gh-pages, or wherever).  Avoiding
the example-side commits also makes sure they don't leak into your
example repository by mistake.  The route for updating the example
repository should be:

1. Your lesson-template PR gets merged
2. The lesson-example maintainers pull from lesson-template
3. The lesson-example maintainers rebuilt the HTML and update the
   central lesson-example repository.

No part of that process needs your example-side commits, so why make
them at all? ;)

[1]: https://github.com/swcarpentry/lesson-example/blob/gh-pages/README.md
[2]: https://github.com/gvwilson/lesson-example/pull/1#discussion_r27498659
[3]: http://en.wikipedia.org/wiki/How-to
[4]: https://github.com/github/hub

karinlag commented on blog/2015/04/how-to-contribute-with-template.html:
```diff
@@ -0,0 +1,112 @@
+---
+<!-- start excerpt -->
+<p><a href="../../2014/09/splitting-the-repo.html">At the end of last year</a> we split out lessons to have one Git repository per topic that we teach. To ensure that all lessons have the same look we create <a href="https://github.com/swcarpentry/lesson-template/">one template for our lessons</a>. During the first quarter of this year we found that have two living branches at the same remote Git ...
+<!-- end excerpt -->
```
I would myself have added one or two sentences about how these two are intended to be used, that way people might gain a clearer understanding of what the functionality of these two repos are.

gvwilson replied:
I agree with @karinlag - we need to make really, really sure people understand that they don't have to do all of this to add to a lesson.

gvwilson commented on blog/2015/04/how-to-contribute-with-template.html:
```diff
@@ -0,0 +1,112 @@
+category: ["How to"]
+---
+<!-- start excerpt -->
+<p><a href="../../2014/09/splitting-the-repo.html">At the end of last year</a> we split out lessons to have one Git repository per topic that we teach. To ensure that all lessons have the same look we create <a href="https://github.com/swcarpentry/lesson-template/">one template for our lessons</a>. During the first quarter of this year we found that have two living branches at the same remote Git ...
```
Please use `{{page.root}}/blog` instead of relative paths.

gvwilson replied:
Please use the variable `github_url` instead of the full URL for our GitHub repos.

gvwilson commented on blog/2015/04/how-to-contribute-with-template.html:
```diff
@@ -0,0 +1,112 @@
+<!-- end excerpt -->
+<h2 id="setup">Setup</h2>
+<p>I start cloning our example</p>
+<pre><code>$ git clone -b gh-pages git@github.com:r-gaia-cs/lesson-example.git</code></pre>
```
Why `<pre><code>` ?  `<pre>` is enough, isn't it?

r-gaia-cs replied:
Because I was lazy. I wrote the blog in Markdown, convert it to HTML using Pandoc and didn't question the output. If you prefer only `<pre>` instead of `<pre><code>` I'm OK in change it.

r-gaia-cs commented:
> * “To ensure that all lessons have the same look we create…” → “To
>   ensure that all lessons have the same look, we created…” (comma and
>   “created”).
> 
> * “… core files of our template
>   <a href="https://github.com/swcarpentry/lesson-template/pull/219">in
>   one repository</a>” should maybe just be linking to lesson-template
>   (not in to the 219 PR)?
> 
> * “Because create a pull request…” → “Because creating a pull
>   request…”
> 
> * “…starting with Git I wrote this how-to” → “starting with Git, I
>   wrote this how-to”.

Out-of-date since I reword this paragraph.

> * I prefer cloning over HTTPS in example docs, see [2] and later.

Fixed.

> * It looks like you're cloning r-gaia-cs/lesson-example.  I expect
>   most folks will want to clone swcarpentry/lesson-example and only
>   fork that repo if they need to update the examples.  If you're just
>   contributing to lesson-template, you don't need a lesson-example
>   fork.

Fixed.

> * I think novice Git users will probably be even more confused with so
>   many remotes for one local repository.  It's probably easier to
>   recommend folks have a separate local repository for each lesson
>   they maintain as well as separate local repositories for
>   lesson-template and lesson-example.  Then each local repository only
>   needs remotes for the central version of that repository and the
>   users personal public fork.  That's less efficient in terms of
>   object de-duplication, but it makes for a simpler mental map.  More
>   advanced Git users will be able to figure out their own preferred
>   remote mapping without our advice.  For template work, you will need
>   an extra remote pointing to a filled-in example repository
>   (e.g. lesson-example), but you won't need two (either
>   r-gaia-cs/swc-lesson-example *or* swcarpentry/lesson-example would
>   be fine).

I dropped `r-gaia-cs/swc-lesson-example`.

AFAIK `git cherry-pick` only work for commits at the current Git repository.
If you know one way to copy one commit
from `lesson-example` to `lesson-template` as separate local repository
that doesn't involved `git format-patch` and `git apply` I will change this.

> * `fit fetch --all` is easier than fetching each remote in turn ;).

I want to avoid add commands/flags that you don't teach.

> * I prefer:
> 
>     $ git merge --ff-only upstream/gh-pages
> 
>   to avoid creating accidental backwards merges.  Alternatively, don't
>   bother updating your local gh-pages and just do:
> 
>     $ git checkout -b example-centering-image upstream/gh-pages
> 
>   for “I'd like to create a feature branch about centering images
>   based on upstream's gh-pages branch”.

Thanks. Addressed.

> * You have both “how-to” and “how to”.  Wikipedia prefers the
>   hyphenated form [3].

Fixed.

> * You have trailing whitespace after “git diff”.

=)

> * “To avoid lose my work I save it at the staged area” → “To avoid
>   losing my work, I stage it in the index”.

I want to avoid the technical wording.

> * “save the change at the HTML file” → “save the change to the HTML
>   file”.

Fixed.

> * “If you attend a Software Carpentry workshop you should be
>   confortable…” → “If you have taken the Git lesson at a Software
>   Carpentry workshop, you should be comfortable…” (specify Git lesson
>   and fix “confortable” typo).

Fixed.

> Instead of committing the template change and HTML to the
> example-based branch, I'd work out the change in the example-based
> branch, then:
> 
> Blow away the side effects:
> 
>   $ git checkout HEAD *.html
> 
> Save my changes for the branch switch:
> 
>   $ git stash
> 
> Switch branches to a feature branch based on the central template:
> 
>   $ git checkout -b center-images swcarpentry-template/gh-pages
> 
> Restore my changes after the branch switch:
> 
>   $ git stash pop
> 
> Commit them:
> 
>   $ git commit -am 'Add img-responsive class to img'
> 
> Publish:
> 
>   $ git push -u my-public-fork-of-lesson-template center-images

I prefer that "novice" do a little more of work and make sure that I can help
them recovery from some problem other than suggest stash than can create big
problems for novice.

> And create a pull request:
> 
>   $ hub pull-request
> 
> (I'd probably follow you and link to the PR docs for the
> instructions, but I thought I'd share my hub [4] use here, because
> it makes my life a lot easier ;).

=)

> The route for updating the example
> repository should be:
> 
> 1. Your lesson-template PR gets merged
> 2. The lesson-example maintainers pull from lesson-template
> 3. The lesson-example maintainers rebuilt the HTML and update the
>    central lesson-example repository.
> 
> No part of that process needs your example-side commits, so why make
> them at all? ;)

Because I'm trying to make it clear for everyone. =)

r-gaia-cs commented:
@gvwilson If you want I can squash the commits before you merge this.

gvwilson commented:
Please don't worry about squashing - it's the end product that matters.  
Otherwise good to go?

r-gaia-cs commented:
> Otherwise good to go?

If the excerpt is good to you, yes. I addressed part of @wking's comments and we can discuss about the others when porting this do `doc` at `lesson-template` or somewhere else.

wking commented:
On Tue, Apr 07, 2015 at 07:44:28AM -0700, Raniere Silva wrote:
> AFAIK `git cherry-pick` only work for commits at the current Git
> repository.  If you know one way to copy one commit from
> `lesson-example` to `lesson-template` as separate local repository
> that doesn't involved `git format-patch` and `git apply` I will
> change this.

No, you're right that you need at least one remote pointing to
lesson-example or another lesson-template consumer.

> > * `fit fetch --all` is easier than fetching each remote in turn
> >   ;).
> 
> I want to avoid add commands/flags that you don't teach.

I don't think we teach fetch at all ;).

  git-novice$ git log --oneline -1
  b21eeea Merge pull request #91 from wking/remove-backup.sh
  git-novice$ git grep fetch
  06-collab.html:origin   https://github.com/vlad/planets.git (fetch)</code></pre>
  06-collab.md:origin   https://github.com/vlad/planets.git (fetch)
  tools/validation_helpers.py:    # Helpers to fetch specific document sections

And I think the pull → fetch transition is a much bigger leap than
‘fetch’ → ‘fetch --all’.

I don't have a big problem listing them all explicitly, but I think
it's reasonable to at least mention the ‘--all’ option in the
surrounding text.

> > * “To avoid lose my work I save it at the staged area” → “To avoid
> >   losing my work, I stage it in the index”.
> 
> I want to avoid the technical wording.

Ok.  The Git lesson currently uses “staging area”, and talks about
“add[ing] the changed files to the staging area” and “put[ing] things
in this area”.  I prefer your “saved” verb, but I like the lesson's
“staging area”.  Git itself also prefers “staging area”:

  git.git$ git grep -i 'staged area' v2.3.4 | wc -l
  0
  git.git$ git grep -i 'staging area' v2.3.4 | wc -l
  46

> > Instead of committing the template change and HTML to the
> > example-based branch, I'd [stuff with ‘git stash’]…
> 
> I prefer that "novice" do a little more of work and make sure that I
> can help them recovery from some problem other than suggest stash
> than can create big problems for novice.

I don't think you can get into serious problems with ‘stash’ (which
keeps the stashed content in the stash stack if there are merge
conflicts while popping it).  And I'm concerned that some of the
example-side commits will leak into the example repository.  But I'm
fine using the cherry-pick approach if the example-side maintainers
are ok asking for (or doing) rebases to drop any such leaked commits.

> > The route for updating the example
> > repository should be:
> > 
> > 1. Your lesson-template PR gets merged
> > 2. The lesson-example maintainers pull from lesson-template
> > 3. The lesson-example maintainers rebuilt the HTML and update the
> >    central lesson-example repository.
> > 
> > No part of that process needs your example-side commits, so why
> > make them at all? ;)
> 
> Because I'm trying to make it clear for everyone. =)

I'm not convinced that the cherry-pick approach is clearer, but I'm
happy to give the cherry-pick approach a spin.

r-gaia-cs commented:
> Git itself also prefers “staging area”:

Fixed.

fmichonneau commented on blog/2015/04/how-to-contribute-with-template.html:
```diff
@@ -0,0 +1,104 @@
+[example-centering-image 1ebf8c3] Update HTML
+ 1 file changed, 2 insertions(+), 1 deletion(-)</code></pre>
+<h2 id="copying-the-change-to-template">Copying the change to template</h2>
+<p>If you have taken the Git lesson at a Software Carpentry workshop, you should be comfortable with what you see until now. What I need to do is copy my change at <code>Makefile</code> to <a href="https://github.com/swcarpentry/lesson-template">template</a> using <code>git cherry-pick</code>. If you are trying to remember it from our workshop, don't panic because we don't teach it but you will d ...
```
discovery -> discover

fmichonneau replied:
The explanation of how to do cherry picking is great, but I think it would be helpful to explain the why: why do you need to cherry pick this particular commit? and also the "what": what does cherry picking do?

fmichonneau commented on blog/2015/04/how-to-contribute-with-template.html:
```diff
@@ -0,0 +1,104 @@
+<pre><code>$ git log --oneline -2
+2004c4f Add img-responsive class to img
+cf25427 Merge pull request #218 from wking/trailing-newlines</code></pre>
+<p>The commit ID change but this is OK and is desired. Now I can upload my changes</p>
```
change -> changes

fmichonneau commented on blog/2015/04/how-to-contribute-with-template.html:
```diff
@@ -0,0 +1,104 @@
+<p>And <a href="https://github.com/swcarpentry/lesson-template">create a pull request</a>, <a href="https://github.com/swcarpentry/lesson-template/pull/219">like this one</a>.</p>
+<h2 id="conclusion">Conclusion</h2>
+<p>Once the pull request was created our contributors <a href="https://github.com/swcarpentry/lesson-template/pull/219#issuecomment-89671094">will review it</a> and eventually merge it. From time to time we will propagate the change at our template to our lessons.</p>
+<p>If your changes are at the CSS that we use you <strong>must</strong> copy our changes to <a href="https://github.com/swcarpentry/styles/">this repository</a> instead of <code>template</code>.</p>
```
at the CSS --> in the CSS

fmichonneau replied:
our changes --> your[?] changes

fmichonneau commented on blog/2015/04/how-to-contribute-with-template.html:
```diff
@@ -0,0 +1,104 @@
+<h2 id="conclusion">Conclusion</h2>
+<p>Once the pull request was created our contributors <a href="https://github.com/swcarpentry/lesson-template/pull/219#issuecomment-89671094">will review it</a> and eventually merge it. From time to time we will propagate the change at our template to our lessons.</p>
+<p>If your changes are at the CSS that we use you <strong>must</strong> copy our changes to <a href="https://github.com/swcarpentry/styles/">this repository</a> instead of <code>template</code>.</p>
+<p>As always, in case of troubles with Git you can contact us and one of your Git gurus will get back to you for help.</p>
```
your Git gurus --> our Git gurus

fmichonneau commented on blog/2015/04/how-to-contribute-with-template.html:
```diff
@@ -0,0 +1,104 @@
+<p>Once the pull request was created our contributors <a href="https://github.com/swcarpentry/lesson-template/pull/219#issuecomment-89671094">will review it</a> and eventually merge it. From time to time we will propagate the change at our template to our lessons.</p>
+<p>If your changes are at the CSS that we use you <strong>must</strong> copy our changes to <a href="https://github.com/swcarpentry/styles/">this repository</a> instead of <code>template</code>.</p>
+<p>As always, in case of troubles with Git you can contact us and one of your Git gurus will get back to you for help.</p>
+<p>P.S.: I take this opportunity to thanks for the help of <a href="https://github.com/fmichonneau">Francois Michonneau</a>, <a href="https://github.com/jpallen">James Allen</a> and <a href="https://github.com/pbanaszkiewicz">Piotr Banaszkiewicz</a> to improve our CSS.</p>
```
you're welcome :)

fmichonneau commented:
Hi @r-gaia-cs 

Sorry for the delay reviewing your post. I think it's really useful.

I see a lot of value as it stands to illustrate how to currently work with the template, and introduce some more advanced features of Git that some instructors may not be familiar with.

I'm also with @wking that this post shouldn't replace documentation, and a summary version on how to work with these repo should also be included with it.

r-gaia-cs commented:
> Sorry for the delay.

No problem. Thanks for doing it.

> I'm also with @wking that this post shouldn't replace documentation,
> and a summary version on how to work with these repo should also be included with it.

https://github.com/wking/swc-lesson-template/pull/1

r-gaia-cs commented:
@gvwilson Can you merge this?

gvwilson commented:
Done - and please feel free to merge others when/as you think they're ready.
